File: fabric-sdk-js/test/crypto.test.js
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import {
  KEY_SIZE,
  NONCE_SIZE,
  TAG_SIZE,
  SALT_SIZE,
  DEFAULT_ARGON2ID_PARAMS,
  seal,
  open,
  randomBytes,
  randomNonce,
  randomSalt,
  deriveKey,
  deriveKeyArgon2id,
} from '../src/crypto.js';

const enc = new TextEncoder();

describe('AEAD (XChaCha20-Poly1305)', () => {
  it('round-trips plaintext', () => {
    const key = randomBytes(KEY_SIZE);
    const nonce = randomNonce();
    const plaintext = enc.encode('Bhai vault payload');
    const aad = enc.encode('namespace=list,stream=001');

    const ct = seal(key, nonce, plaintext, aad);
    assert.equal(ct.length, plaintext.length + TAG_SIZE);

    const pt = open(key, nonce, ct, aad);
    assert.deepEqual(pt, plaintext);
  });

  it('rejects wrong key', () => {
    const k1 = randomBytes(KEY_SIZE);
    const k2 = randomBytes(KEY_SIZE);
    const nonce = randomNonce();
    const ct = seal(k1, nonce, enc.encode('x'), null);
    assert.throws(() => open(k2, nonce, ct, null));
  });

  it('rejects wrong nonce', () => {
    const key = randomBytes(KEY_SIZE);
    const ct = seal(key, randomNonce(), enc.encode('x'), null);
    assert.throws(() => open(key, randomNonce(), ct, null));
  });

  it('rejects tampered ciphertext', () => {
    const key = randomBytes(KEY_SIZE);
    const nonce = randomNonce();
    const ct = seal(key, nonce, enc.encode('hello'), null);
    ct[0] ^= 0x01;
    assert.throws(() => open(key, nonce, ct, null));
  });

  it('rejects wrong AAD', () => {
    const key = randomBytes(KEY_SIZE);
    const nonce = randomNonce();
    const ct = seal(key, nonce, enc.encode('hello'), enc.encode('aad-a'));
    assert.throws(() => open(key, nonce, ct, enc.encode('aad-b')));
  });

  it('rejects bad key length', () => {
    assert.throws(() => seal(new Uint8Array(16), randomNonce(), new Uint8Array(), null));
  });

  it('rejects bad nonce length', () => {
    assert.throws(() => seal(randomBytes(KEY_SIZE), new Uint8Array(12), new Uint8Array(), null));
  });
});

describe('Random sources', () => {
  it('randomNonce is unique across draws', () => {
    const seen = new Set();
    for (let i = 0; i < 64; i++) {
      const n = randomNonce();
      assert.equal(n.length, NONCE_SIZE);
      const k = Buffer.from(n).toString('hex');
      assert.equal(seen.has(k), false, 'duplicate nonce — random source broken');
      seen.add(k);
    }
  });

  it('randomSalt returns 16 bytes', () => {
    assert.equal(randomSalt().length, SALT_SIZE);
  });
});

describe('HKDF-SHA256', () => {
  it('is deterministic', async () => {
    const secret = enc.encode('master-key-material');
    const salt = enc.encode('salt-bytes');
    const info = enc.encode('namespace=list');
    const a = await deriveKey(secret, salt, info);
    const b = await deriveKey(secret, salt, info);
    assert.deepEqual(a, b);
    assert.equal(a.length, KEY_SIZE);
  });

  it('different info → different key', async () => {
    const secret = enc.encode('master-key-material');
    const salt = enc.encode('salt-bytes');
    const a = await deriveKey(secret, salt, enc.encode('namespace=list'));
    const b = await deriveKey(secret, salt, enc.encode('namespace=bahi'));
    assert.notDeepEqual(a, b);
  });

  it('different salt → different key', async () => {
    const secret = enc.encode('master-key-material');
    const info = enc.encode('namespace=list');
    const a = await deriveKey(secret, enc.encode('salt-a'), info);
    const b = await deriveKey(secret, enc.encode('salt-b'), info);
    assert.notDeepEqual(a, b);
  });
});

describe('Argon2id', () => {
  it('is deterministic', async () => {
    const salt = enc.encode('0123456789abcdef');
    const a = await deriveKeyArgon2id('correct horse battery staple', salt);
    const b = await deriveKeyArgon2id('correct horse battery staple', salt);
    assert.deepEqual(a, b);
    assert.equal(a.length, KEY_SIZE);
  });

  it('different passphrase → different key', async () => {
    const salt = enc.encode('0123456789abcdef');
    const a = await deriveKeyArgon2id('pass-a', salt);
    const b = await deriveKeyArgon2id('pass-b', salt);
    assert.notDeepEqual(a, b);
  });

  it('default params match spec (t=3, m=65536, p=4, hash=32)', () => {
    assert.equal(DEFAULT_ARGON2ID_PARAMS.time, 3);
    assert.equal(DEFAULT_ARGON2ID_PARAMS.memory, 65536);
    assert.equal(DEFAULT_ARGON2ID_PARAMS.parallelism, 4);
    assert.equal(DEFAULT_ARGON2ID_PARAMS.hashLength, KEY_SIZE);
  });
});

File: fabric-sdk-go/crypto/crypto_test.go
package crypto

import (
	"bytes"
	"testing"
)

func mustKey(t *testing.T) []byte {
	t.Helper()
	k, err := RandomBytes(KeySize)
	if err != nil {
		t.Fatalf("RandomBytes: %v", err)
	}
	return k
}

func mustNonce(t *testing.T) []byte {
	t.Helper()
	n, err := RandomNonce()
	if err != nil {
		t.Fatalf("RandomNonce: %v", err)
	}
	return n
}

func TestAEADRoundTrip(t *testing.T) {
	key := mustKey(t)
	nonce := mustNonce(t)
	plaintext := []byte("Bhai's vault payload")
	aad := []byte("namespace=list,stream=001")

	ct, err := Seal(key, nonce, plaintext, aad)
	if err != nil {
		t.Fatalf("Seal: %v", err)
	}
	if len(ct) != len(plaintext)+TagSize {
		t.Fatalf("ciphertext length: got %d, want %d", len(ct), len(plaintext)+TagSize)
	}

	got, err := Open(key, nonce, ct, aad)
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	if !bytes.Equal(got, plaintext) {
		t.Fatalf("plaintext mismatch: got %q, want %q", got, plaintext)
	}
}

func TestAEADWrongKeyFails(t *testing.T) {
	key1 := mustKey(t)
	key2 := mustKey(t)
	nonce := mustNonce(t)
	ct, err := Seal(key1, nonce, []byte("x"), nil)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := Open(key2, nonce, ct, nil); err == nil {
		t.Fatal("Open with wrong key should fail")
	}
}

func TestAEADWrongNonceFails(t *testing.T) {
	key := mustKey(t)
	n1 := mustNonce(t)
	n2 := mustNonce(t)
	ct, err := Seal(key, n1, []byte("x"), nil)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := Open(key, n2, ct, nil); err == nil {
		t.Fatal("Open with wrong nonce should fail")
	}
}

func TestAEADTamperedCiphertextFails(t *testing.T) {
	key := mustKey(t)
	nonce := mustNonce(t)
	ct, err := Seal(key, nonce, []byte("hello"), nil)
	if err != nil {
		t.Fatal(err)
	}
	ct[0] ^= 0x01
	if _, err := Open(key, nonce, ct, nil); err == nil {
		t.Fatal("Open with tampered ciphertext should fail")
	}
}

func TestAEADWrongAADFails(t *testing.T) {
	key := mustKey(t)
	nonce := mustNonce(t)
	ct, err := Seal(key, nonce, []byte("hello"), []byte("aad-a"))
	if err != nil {
		t.Fatal(err)
	}
	if _, err := Open(key, nonce, ct, []byte("aad-b")); err == nil {
		t.Fatal("Open with wrong AAD should fail")
	}
}

func TestSealRejectsBadKeyAndNonce(t *testing.T) {
	if _, err := Seal(make([]byte, 16), make([]byte, NonceSize), nil, nil); err != ErrInvalidKeySize {
		t.Errorf("short key: got %v, want ErrInvalidKeySize", err)
	}
	if _, err := Seal(make([]byte, KeySize), make([]byte, 12), nil, nil); err != ErrInvalidNonceSize {
		t.Errorf("short nonce: got %v, want ErrInvalidNonceSize", err)
	}
}

func TestRandomNonceUniqueness(t *testing.T) {
	seen := map[string]struct{}{}
	for i := 0; i < 64; i++ {
		n, err := RandomNonce()
		if err != nil {
			t.Fatal(err)
		}
		if len(n) != NonceSize {
			t.Fatalf("len=%d, want %d", len(n), NonceSize)
		}
		if _, dup := seen[string(n)]; dup {
			t.Fatal("duplicate nonce — random source broken")
		}
		seen[string(n)] = struct{}{}
	}
}
